### DISCUSSION TITLE
How should have ellipsis in button

### DISCUSSION THREAD

**Discussion by Monirzadeh** [koreader/koreader#12201] [Q&A] [answered]
i am working on a plugin in some place i have an specific button like this
```
	local word_widget = Button:new({
		text = self.item.word,
		bordersize = 0,
		callback = function()
			-- some function call here
		end,
		padding = 0,
	})
```
if i set `width` or `max_width` for this butons in long text it split text in two line with smaller font.
i want it keep the size and ellipsis that(similar to table)
so somthing like
```
a really long text 
split in two line
```
show  like this
```
a really long t...
```
how should i done that?
Note: set style like this will not affect if i set  `width` or `max_width` 
```
word_widget.style = "white-space:nowrap; overflow:hidden; text-overflow:ellipsis"
```
if i set that without `width` or `max_width` than box will get out of screen

**Comment by poire-z** [comment]
> `word_widget.style = "white-space:nowrap; overflow:hidden; text-overflow:ellipsis"`

KOReader UI is not like most of the fancy UI stuff made after 1998: it doesn't support styling with CSS, it is old school.

Most of the options (and their default values) a widget supports are documented in the widget `:init()` or `:extend()` , ie. for Button:
https://github.com/koreader/koreader/blob/5affb68233fda994150df1da3d0b573402f20842/frontend/ui/widget/button.lua#L37-L69

So, I guess you could get what you want by providing `avoid_text_truncation = false` (as the default is `true`).

**Reply by Monirzadeh** [reply]
Thanks :heart:
is it text inside that get auto direction? RTL text get `align` right and LTR get `align` left automatically?

Update: i test that and it still split line with `width`

**Reply by Monirzadeh** [reply]
as you see it ellipsis but split line too. use `width` not `max_width`
![test](https://github.com/user-attachments/assets/e23efe5b-dca4-44cd-90d7-14c04f0435fa)

**Reply by benoit-pierre** [reply]
There's also `height_overflow_show_ellipsis`, I think.

**Reply by Monirzadeh** [reply]
can you please share an example.
```	local word_widget = Button:new({
		text = self.item.word,
		bordersize = 0,
		callback = function()
			-- some function call here
		end,
		padding = 0,
                width = 700,
               height_overflow_show_ellipsis = true,
	})
```
or put that on `:extend()` not working.
should i use that somewhere else?

**Reply by benoit-pierre** [reply, marked as answer]
700 is a lot, no? Anyway, this seem to do the right thing:
```lua
    self.star = Button:new{
        text = "a really long text\nsplit in two line",
        avoid_text_truncation = false,
        width = 200,
        enabled = button_enabled,
        show_parent = self,
        readonly = self.readonly,
    }
```
Mangling `frontend/ui/widget/bookstatuswidget.lua` to test:
![2024-07-22-012652_481x195_scrot](https://github.com/user-attachments/assets/6362e74a-c831-4338-8b9a-5f099d03b6f3)

**Reply by Monirzadeh** [reply]
Thanks everybody it is working.
is it possible alignment Right or left based on content of that instead of manually `align`?

i something similar to a table with multiple widget inside that.
i will make that tittle clickable with other widget around that than putt them in a table. it is why it is so long

**Reply by poire-z** [reply]
If you use `align = "left"`, it should auto-do "right" if your text or the UI lanuage is RTL.
Isn't that the case?

(Not for you to touch, but the TextBoxWidget used inside a Button has:
https://github.com/koreader/koreader/blob/5affb68233fda994150df1da3d0b573402f20842/frontend/ui/widget/textboxwidget.lua#L104-L105
)

**Reply by Monirzadeh** [reply]
the problem is this option affect with text and UI both
so if UI be RTL and text be LTR it align RTL (because of UI)
i need regardless of UI align text automatically. get RTL if text is RTL get LTR if text is RTL

**Reply by poire-z** [reply]
Ok. We don't forward other parameters from Button to Text*Widget, because we never needed that use case.
May be you can try this hack:
```lua
button = Button:new{ text=your_text, ... your button as it was }
button.label_widget.auto_para_direction = true
button.label_widget:setText("")
button.label_widget:setText(your_text)
```

**Reply by Monirzadeh** [reply]
i test this and align center :(

**Reply by poire-z** [reply]
You would still need to specify `align = "left"`, did you?

**Reply by Monirzadeh** [reply]
it is the exact code
```
	local word_widget = Button:new({
		text = self.item.word,
		avoid_text_truncation = false,
		width = self.width,
		bordersize = 0,
		align = "left",
		callback = function()
			self.item.callback(self.item)
		end,
		padding = 0,
	})
	word_widget.label_widget.auto_para_direction = true
	word_widget.label_widget:setText("")
	word_widget.label_widget:setText(self.item.word)
	word_widget.label_widget.fgcolor = self.item.is_dim and dim_color or Blitbuffer.COLOR_BLACK
```

**Reply by poire-z** [reply]
Well, I tried that with the buttons of our SkimWidget, and indeed it doesn't work.
The fact is that we use a TextWidget is auto sized to its content when using `avoid_text_truncation = false`, instead of a TextBoxWidget (which can have a width larger than its content, so that the content can be auto aligned left or right).
I can't make out a quick hack :/ You'd need to create dedicated widget inspired by Button, using only TextBoxWidget, which may be very complicated. Sorry :/

**Reply by Monirzadeh** [reply]
thanks for investigation. it is get me a more clear view.
if i can't handle that with other alternative will try that :+1:
